### PR TITLE
Add BC mechanism to Module.load_state_dict

### PR DESCRIPTION
**Pull request by ssnl**
1. Add version number to module. Currently all modules have version `0`. Modules without version numbers are considered as having `-1` as their versions.
2. Change `load_state_dict` to call `load_local_state_dict` on each submodule. This method can be overridden for class specific loading.
3. Change `InstanceNorm*d`'s hard error (when `track_running_stats=False` and `running* in state_dict`) to a warning because in most cases the running buffers are not used and manually removing the keys is quite some trouble for users.

Unblocks #6445 .

### CODE REVIEW

**Comment by SsnL** [comment]
onnx-fb-universe can be fixed by https://github.com/onnxbot/onnx-fb-universe/pull/1601 . We can merge that if someone thinks that this is good.

**Comment by SsnL** [comment]
Let me know if I should just make the version number of LongTensot scalar. It makes other code easier..

**Comment by SsnL** [comment]
@apaszke Having a global counter also works for me. I remember @soumith has some opinion against it.

However, when the `_version` is missing, shouldn't we assume that it is the old version, i.e. before 0.4?

**Comment by apaszke** [comment]
I'm happy to discuss that if there are any concerns. The reason why we should assume is the newest version, is that someone might actually construct the state dict (half-)manually, e.g. by stitching together a few other dicts. In that case they will lack the field, but the best thing we can do is assume that they are matching our new formats, not the legacy ones.

**Comment by SsnL** [comment]
@apaszke I agree that we shouldn't add entries to `state_dict`. However, adding module-specific version counters makes maintaining easier, and people will less likely to forget incrementing it. 

Therefore, I add an attr `state_dict._versions` specifically for storing version counters. Let me know what you think!

File: torch/jit/__init__.py [hunk 1]
```diff
@@ -661,6 +671,9 @@ def __getitem__(self, k):
             raise KeyError(k)
         return self.module._get_parameter(k)
 
+    def __len__(self):
+        return len(self.items())
```
Comment (apaszke) [inline torch/jit/__init__.py]: Why do we need this now?

**Comment by apaszke** [comment]
@SsnL I'll review it today, but you have to agree that the code is significantly more complicated now 😕

File: torch/nn/modules/instancenorm.py [hunk 2]
```diff
@@ -11,22 +13,27 @@ def __init__(self, num_features, eps=1e-5, momentum=0.1, affine=False,
...
-                'anymore since 0.4.0. If running stats are not used, remove '
-                'them from state_dict before calling load_state_dict. '
-                'Otherwise, set track_running_stats=True in {klass} to use '
-                'running stats. See the documentation of {klass} for more '
-                'details.'
-                .format(name=full_name, klass=self.__class__.__name__))
-        super(_InstanceNorm, self)._load_state_dict_key_mismatch(
-            full_name, name, is_missing)
+    def load_local_state_dict(self, local_state_dict, version, strict):
+        # at version 0: removed running_mean and running_var when
```
Comment (apaszke) [inline torch/nn/modules/instancenorm.py]: Can we make this 1? Otherwise it means that previously all modules had version -1, which is a bit counterintuitive.

File: torch/nn/modules/instancenorm.py [hunk 3]
```diff
@@ -11,22 +13,27 @@ def __init__(self, num_features, eps=1e-5, momentum=0.1, affine=False,
...
+                    'Unexpected running stats buffer(s) {names} in state_dict '
+                    'for {klass} with track_running_stats=False. If you are '
+                    'trying to load a checkpoint saved before 0.4.0, this may '
+                    'be expected because {klass} does not track running stats '
+                    'by default anymore since 0.4.0. These buffers won\'t be '
+                    'actually loaded. If these stats are actually needed, set '
+                    'track_running_stats=True in {klass} to enable running '
+                    'stats. See the documentation of {klass} for details.'
+                    .format(names=" and ".join('"{}"'.format(k) for k in running_stats_keys),
+                            klass=self.__class__.__name__))
```
Comment (apaszke) [inline torch/nn/modules/instancenorm.py]: Why is that a warning now? It should really be an error, and someone should either patch their state dict, or change the flag to the constructor of this module

File: torch/nn/modules/module.py [hunk 4]
```diff
@@ -47,6 +47,30 @@ def forward(self, x):
 
     dump_patches = False
 
+    r"""This allows better BC support for :meth:`load_state_dict`. In
+    :meth:`state_dict`, the version number will be saved as in the attribute
+    `_versions` of the returned state dict, and thus pickled. `_versions` is a
+    dictionary. Its keys follow the naming convention of state dict, but ends
+    with key ``"version"`` and is associated with the version number of that
```
Comment (apaszke) [inline torch/nn/modules/module.py]: nit: we could probably skip the `.version` suffix because it's unnecessary. Otherwise, if we anticipate that we might want to stash some extra metadata in the future, then we could rename `._versions` to `._metadata`

File: torch/nn/modules/module.py [hunk 5]
```diff
@@ -566,13 +590,16 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
...
-                if not None, the return dictionary is stored into destination.
-                Default: None
+                if not ``None``, the return dictionary is stored into
+                :attr:`destination`. Default: None
             prefix (string, optional): Adds a prefix to the key (name) of every
                 parameter and buffer in the result dictionary. Default: ''
             keep_vars (bool, optional): if ``True``, returns a Variable for each
                 parameter. If ``False``, returns a Tensor for each parameter.
                 Default: ``False``
+            no_child (bool, optional): if ``True``, excludes parameters and
```
Comment (apaszke) [inline torch/nn/modules/module.py]: Why?

File: torch/nn/modules/module.py [hunk 6]
```diff
@@ -586,87 +613,162 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
         """
         if destination is None:
             destination = OrderedDict()
+            destination._versions = OrderedDict()
+        if hasattr(destination, '_versions'):
```
Comment (apaszke) [inline torch/nn/modules/module.py]: Why would it ever not have this attribute here? We should really make all of those args private, they're not meant to be used by others

File: torch/nn/modules/module.py [hunk 7]
```diff
@@ -586,87 +613,162 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
...
 
-    def _load_state_dict_key_mismatch(self, full_name, name, is_missing):
-        r"""This is called in :meth:`~torch.nn.Module.load_state_dict` when
-        there is state dict key mismatch in ``strict=True`` mode. This method
-        can be overridden by subclasses to raise class-specific errors.
-
-        When :attr:`is_missing` is ``True``, :attr:`full_name` can not be found in
-        the dict being loaded. When :attr:`is_missing` is ``False``,
-        :attr:`full_name` is unexpected in the dict being loaded.
+    def load_local_state_dict(self, local_state_dict, version, strict):
```
Comment (apaszke) [inline torch/nn/modules/module.py]: Can we call it `load_from_state_dict`? I'd like to avoid having multiple different concepts called `state_dict`.

Comment (apaszke) [reply torch/nn/modules/module.py]: I'm not sure if we want to add this to the API just yet. E.g. `_version` is a private attribute, but it is necessary if someone would like to use it. On the other hand, making every `nn.Module` have a `.version` attr doesn't seem nice. Let's just keep it private so it can get into the release quickly, and later we can think of ways to clean it up

File: torch/nn/modules/module.py [hunk 8]
```diff
@@ -586,87 +613,162 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
...
-        def submodule_key_mismatch(full_name, is_missing):
-            module = self
-            names = full_name.split(".")
-            for module_name in names[:-1]:
-                if module_name in module._modules:
-                    module = module._modules[module_name]
-                else:
-                    return
-            module._load_state_dict_key_mismatch(full_name, names[-1], is_missing)
+        own_local_state = self.state_dict(no_child=True)
```
Comment (apaszke) [inline torch/nn/modules/module.py]: Do we really need `no_child`? `own_local_state` is basically this:
```python
state = {k: v.data for k, v in chain(self._parameters.items(), self._buffers.items())}
```

File: torch/nn/modules/module.py [hunk 9]
```diff
@@ -586,87 +613,162 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
...
-                submodule_key_mismatch(name, True)
+            missing = set(own_local_state.keys()) - set(local_state_dict.keys())
             error_msg = ''
             if len(unexpected) > 0:
                 error_msg += 'Unexpected key(s) in state_dict: {}. '.format(
                     ', '.join('"{}"'.format(k) for k in unexpected))
             if len(missing) > 0:
                 error_msg += 'Missing key(s) in state_dict: {}. '.format(
-                    ', '.join('"{}"'.format(k) for k in unexpected))
+                    ', '.join('"{}"'.format(k) for k in missing))
```
Comment (apaszke) [inline torch/nn/modules/module.py]: Do the keys have prefix removed at this point? If so, then this error will be very uninformative.

**Review by apaszke — CHANGES_REQUESTED** [review]
I don't see how the addition of `version` requires us to implement this whole tree logic for modules. I don't think it's worth it, and I certainly don't feel comfortable merging this in the day of a release.

I think that if we removed the `.version` suffix from `._versions`, then we could easily look up the version for a module just by slicing out the param name. Can we please instead:
- Keep the old implementation for `load_state_dict`
- Make `_load_from_state_dict` private
- Make `_load_from_state_dict` take the input state dict, thing returned from `state_dict()` in `load_state_dict`, a prefix of this module, and load from this (basically you'd paste the default body of the loop and modify it slightly).
- make it append missing/unexpected keys to sets given as arguments. Error reporting for key mismatches should be a responsibility of the base method, unless it's really a very specific thing.

I'm happy to coordinate and try to get this in today, feel free to ping me on the chat if you have any questions.

File: torch/nn/modules/instancenorm.py (hunk 3 above)
Comment (SsnL) [reply torch/nn/modules/instancenorm.py]: The more I worked on tutorials code (and my research code), the more I want this to be a warning. So many people will try using community repo code to load existing checkpoints trained with InstanceNorm, and we can't rely on the repos being updated. From the users perspective, they may not be familiar with how state_dict works, or even not very familiar with PyTorch, but just want to run a cool model like neural style transfer. So I really really want to have this as warning.

**Comment by SsnL** [comment]
@apaszke I address the comments. However, indexing into the `_metadata` (or `_versions`) with `prefix` is kinda not very nice.. For root `prefix=""`. For its direct child `prefix="linear."`. All submodules will have the trailing dot, but we have to live with it since apparently `module.add_module('', nn.Conv2d(1,1,1))` works. :/

I made it `_metadata` in case we want to add more stuff in future. 

On the InstanceNorm warning, I replied to you above.

**Comment by SsnL** [comment]
I just realized that `x.add_module('.', nn.Conv1d(1,1,1))` also works... So our current state_dict is ambiguous itself! Should I add a check in `__setattr__` for `"." in key`?

File: torch/nn/modules/instancenorm.py (hunk 3 above)
Comment (apaszke) [reply torch/nn/modules/instancenorm.py]: I'm a bit worried about this. It is completely inconsistent with our `strict` setting, which means it basically violates the contract of `load_state_dict`. Will we always want to make those things warnings?

Your argument make sense, but this can really be fixed extremely easily (`strict=False`), so it's not that convincing for me.

File: torch/nn/modules/module.py [hunk 10]
```diff
@@ -564,11 +583,6 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
         (rather than a Tensor).
 
         Args:
-            destination (dict, optional):
-                if not None, the return dictionary is stored into destination.
-                Default: None
-            prefix (string, optional): Adds a prefix to the key (name) of every
-                parameter and buffer in the result dictionary. Default: ''
             keep_vars (bool, optional): if ``True``, returns a Tensor for each
```
Comment (apaszke) [inline torch/nn/modules/module.py]: This one should be private too. The message is really funny now that we don't have Variables 😄

File: torch/nn/modules/module.py [hunk 11]
```diff
@@ -583,8 +597,18 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
             ['bias', 'weight']
 
         """
-        if destination is None:
-            destination = OrderedDict()
+        if destination is not None:
+            raise ValueError("destination argument is deprecated")
+        if prefix is not '':
+            raise ValueError("prefix argument is deprecated")
```
Comment (apaszke) [inline torch/nn/modules/module.py]: Oh that's not what I meant, sorry if I was unclear. :confused: I only wanted to remove the docs for those functions, no need to implement a private version of this method.

File: torch/nn/modules/module.py [hunk 12]
```diff
@@ -593,79 +617,103 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
...
+                                       .format(key, param.size(), input_param.size()))
             elif strict:
-                unexpected.append(name)
+                missing_keys.append(key)
+
+        if strict:
+            for key, input_param in state_dict.items():
+                if key.startswith(prefix):
+                    input_name = key[len(prefix):]
+                    if '.' not in input_name and input_name not in local_state:
```
Comment (apaszke) [inline torch/nn/modules/module.py]: Why is the `'.' not in input_name` condition here? I'd say that having a dot at this point would be highly unexpected and shouldn't cause us to *not* report an error

File: torch/nn/modules/module.py [hunk 13]
```diff
@@ -593,79 +617,103 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
...
+        metadata = state_dict._metadata if hasattr(state_dict, '_metadata') else None
+        state_dict = state_dict.copy()
+        if metadata is not None:
+            state_dict._metadata = metadata
+
+        def load(module, prefix=''):
+            module._load_from_state_dict(state_dict, prefix, strict, missing_keys, unexpected_keys)
+            for name, child in module._modules.items():
+                if child is not None:
+                    load(child, prefix + name + '.')
```
Comment (apaszke) [inline torch/nn/modules/module.py]: We should maintain a per-module memo here. Otherwise you can end up with multiple modules getting called multiple times, and because e.g. `missing_keys` is not a `set`, then some keys can be reported as missing multiple times. We don't have an invariant saying that modules form a tree. They can be a general DAG or regular graph.

**Review by apaszke — COMMENTED** [review]
That looks much better and is way simpler, thanks! There are a few things, but they're mostly small. We should discuss the InstanceNorm issue.

I also think there's an extra argument for having a global version: it makes it fairly easy for users to actually fix the loading of state dict that somehow ended up without a metadata, by specifying a single argument to `load_state_dict`. Wdyt?

File: torch/nn/modules/module.py (hunk 12 above)
Comment (SsnL) [reply torch/nn/modules/module.py]: Because the `key` might be a key for a param in a child of `this` :)

File: torch/nn/modules/module.py (hunk 13 above)
Comment (SsnL) [reply torch/nn/modules/module.py]: Right.. but if it is a dag like below:
```
  A 
 /   \
B   C
 \   /
  D

all arrows from top to bottom
```
Then `D`'s param will still occur twice in `A.state_dict()` each time with different keys, right?

File: torch/nn/modules/instancenorm.py (hunk 3 above)
Comment (SsnL) [reply torch/nn/modules/instancenorm.py]: The original IN behavior is really subtle. It only uses running stats when `.train(False)`, but not `.eval()`. And using running stats is a non-standard thing to do with IN.

 So I think among people that loads a `state_dict` with `running_*`, ones who actually wants to use them are extremely few. Since we don't want to advocate using `strict=False` as it suppresses other mismatches, I think it is reasonable to break the strict protocol here a bit, and print the warning.

File: torch/nn/modules/module.py (hunk 11 above)
Comment (SsnL) [reply torch/nn/modules/module.py]: I reverted this part in new commit. I did the deprecation because the `._metadata[prefix] = ...` thing will error if people actually used `destination` argument. But I think it should be fine as probably no one uses it.

File: test/test_nn.py [hunk 14]
```diff
@@ -2606,7 +2620,7 @@ def test_load_state_dict(self):
         del old_state_dict['block.conv1.bias']
         del old_state_dict['bn.running_mean']
         for k, v, in old_state_dict.items():
-            self.assertTrue(v.equal(new_state_dict[k]))
+            self.assertEqual(v, new_state_dict[k])
```
Comment (apaszke) [inline test/test_nn.py]: Please set precision to `0` here. `.equal` is more strict

File: torch/nn/modules/instancenorm.py [hunk 15]
```diff
@@ -11,22 +11,36 @@ def __init__(self, num_features, eps=1e-5, momentum=0.1, affine=False,
...
-                'them from state_dict before calling load_state_dict. '
-                'Otherwise, set track_running_stats=True in {klass} to use '
-                'running stats. See the documentation of {klass} for more '
-                'details.'
-                .format(name=full_name, klass=self.__class__.__name__))
-        super(_InstanceNorm, self)._load_state_dict_key_mismatch(
-            full_name, name, is_missing)
+    def _load_from_state_dict(self, state_dict, prefix, strict, missing_keys, unexpected_keys, error_msgs):
+        try:
+            version = state_dict._metadata[prefix]["version"]
```
Comment (apaszke) [inline torch/nn/modules/instancenorm.py]: So we do use those weird keys now? Can we change them to `prefix[:-1]`? This will be an empty string for an empty string, and will remove the trailing dot otherwise.

File: torch/nn/modules/module.py [hunk 16]
```diff
@@ -596,76 +604,109 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
...
+                                      'whose dimensions in the checkpoint are {}.'
+                                      .format(key, param.size(), input_param.size()))
+            elif strict:
+                missing_keys.append(key)
+
+        if strict:
+            for key, input_param in state_dict.items():
+                if key.startswith(prefix):
+                    input_name = key[len(prefix):]
+                    input_name = input_name.split('.', 1)[0]  # get the name of param/buffer/child
```
Comment (apaszke) [inline torch/nn/modules/module.py]: Why is this necessary? Wouldn't it be enough to check that there's no `.` in `input_name`?

File: torch/jit/__init__.py [hunk 17]
```diff
@@ -774,9 +775,10 @@ def _get_methods(cls):
 _compiled_methods_whitelist = {
     'forward', 'register_buffer', 'register_parameter', 'add_module',
     '_apply', 'apply', 'cuda', 'cpu', 'type', 'float', 'double', 'half',
-    'state_dict', 'load_state_dict', 'parameters', 'named_parameters',
-    '_all_buffers', 'children', 'named_children', 'modules', 'named_modules',
-    'zero_grad', 'share_memory', '_get_name'
+    'state_dict', '_state_dict', 'load_state_dict', '_load_from_state_dict',
```
Comment (apaszke) [inline torch/jit/__init__.py]: Can you please remove `_state_dict`?

File: torch/nn/modules/instancenorm.py (hunk 15 above)
Comment (SsnL) [reply torch/nn/modules/instancenorm.py]: We have to use these weird keys, because you can do `m.add_module('', submodule)`

File: torch/nn/modules/module.py (hunk 16 above)
Comment (SsnL) [reply torch/nn/modules/module.py]: This is necessary. If we just check for `.`, then we won't detect the submodules with names not in `self._modules`

File: torch/nn/modules/instancenorm.py (hunk 15 above)
Comment (apaszke) [reply torch/nn/modules/instancenorm.py]: What does that change? It will still slice out the last dot, I can't see the problem

**Review by apaszke — COMMENTED** [review]
Let's please remove the dot at the end of `_metadata` keys and this is good to land

**Review by apaszke — APPROVED** [review]

**Comment by gchanan** [comment]
failures don't look related.

File: torch/nn/modules/module.py [hunk 18]
```diff
@@ -142,8 +163,12 @@ def add_module(self, name, module):
         if not isinstance(module, Module) and module is not None:
             raise TypeError("{} is not a Module subclass".format(
                 torch.typename(module)))
-        if hasattr(self, name) and name not in self._modules:
+        elif hasattr(self, name) and name not in self._modules:
             raise KeyError("attribute '{}' already exists".format(name))
+        elif '.' in name:
+            raise KeyError("module name can't contain \".\"")
```
Comment (lanpa) [inline torch/nn/modules/module.py]: This breaks the densenet's implementation in torchvision. See the error at https://travis-ci.org/lanpa/tensorboard-pytorch/jobs/369432497

File: torch/nn/modules/module.py [hunk 19]
```diff
@@ -597,76 +611,109 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
...
+        unexpected_keys = []
+        error_msgs = []
+
+        # copy state_dict so _load_from_state_dict can modify it
+        metadata = getattr(state_dict, '_metadata', None)
+        state_dict = state_dict.copy()
+        if metadata is not None:
+            state_dict._metadata = metadata
+
+        def load(module, prefix=''):
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/nn/modules/module.py]: `load` here does not necessarily have to be a local func def. However, that `load_state_dict` is not a recursive call predates this patch. If I were to guess, I would say that this probably made the original implementation easier. For custom modules, you can, for now, overwrite `_load_from_state_dict` instead. However, note that it is a private API at the moment.

[the remark this answers is not in the document]
Comment (mahnerak) [reply torch/nn/modules/module.py]: @SsnL Thanks for your quick response.
The problem is that by overriding `_load_from_state_dict` you can't change the behaviour of how you treat children modules.
Say my custom Module has certain children Modules and I don't want to serialize/deserialize the values of the parameters of all the models (save only modules those are not blacklisted in that custom module). Here I can't override the behaviour of traversal, as it is fixed to the top module.
It would be great to have the ability to override the recursive traversal code.
The potential usage is saving/loading of very compact models on top of huge pretrained models. By having control over the loading behaviour we can drastically reduce footprint of checkpoints.

[the remark this answers is not in the document]
Comment (mahnerak) [reply torch/nn/modules/module.py]: @SsnL May I consider creating Pull Request to make `load` a method of `Module`?
